### ISSUE TITLE
protontricks doesn't recognise case-insensitive filesystems

### ISSUE THREAD

**Issue by craig-sanders** [Matoking/protontricks#112]
`protontricks` complains about the existence of both "steamapps" and "SteamApps" directories on **case-insensitive** file-systems  (where only one such directory actually exists).  My guess is that it does this because when it tests for the existence of both variations of the directory name, it will get a true result for both - because they're the exact same directory on a case-insensitive filesystem.

Protontricks should at least compare the inode numbers of the paths to see if they're the same before warning about a non-existent problem.  If both variants "exist" and the inodes are different, then that's a real problem that needs a warning message.  If both exist and the inodes are the same, it's not a real problem, it's a case-insensitive fs (at least for Linux filesystems like ZFS, JFS, and ext4 - all of which have optional support for case-insensitivity.  A different test may be needed for NTFS & FAT).

Example of the problem:

```
$ protontricks -s dishonored
protontricks (WARNING): Both 'steamapps' and 'SteamApps' directories were found at /var/games/steam2. 'SteamApps' directory should be removed to prevent issues with app and Proton discovery.
protontricks (WARNING): Both 'steamapps' and 'SteamApps' directories were found at /var/games/ssd. 'SteamApps' directory should be removed to prevent issues with app and Proton discovery.
Found the following games:
Dishonored (205100)
Dishonored 2 (403640)
Dishonored®: Death of the Outsider™  (614570)

To run protontricks for the chosen game, run:
$ protontricks APPID COMMAND

NOTE: A game must be launched at least once before protontricks can find the game.
```

Saying `'SteamApps' directory should be removed to prevent issues with app and Proton discovery.` is completely wrong for a case-insensitive filesystem, and may cause naive users to delete their steam library.

On this system, /var/games/steam2 and /var/games/ssd are ZFS datasets with case-sensitivity disabled (necessary because lots of windows devs are sloppy and careless about upper & lower case in file and directory names...on a case-sensitive fs, the only way to fix is to create a symlink for each variant spelling).  One is on an ssd pool ("hex"), the other is on a pool of hdds ("exp").  BTW, I also have separate datasets on the same pools for my linux native games - they're case-**sensitive**, as they should be.

Anyway, the following shows that they're the same directory name on case-insensitive filesystems:

```
$ df -Th /var/games/steam2 /var/games/ssd 
Filesystem        Type  Size  Used Avail Use% Mounted on
exp/games/steam2  zfs   9.4T  3.0T  6.5T  31% /var/games/steam2
hex/var/games/ssd zfs   265G  128K  265G   1% /var/games/ssd

$ zfs get casesensitivity hex/var/games/ssd exp/games/steam2
NAME               PROPERTY         VALUE        SOURCE
exp/games/steam2   casesensitivity  insensitive  -
hex/var/games/ssd  casesensitivity  insensitive  -


$ stat -c '%i %n' /var/games/steam2/steamapps/ /var/games/steam2/SteamApps/
2 /var/games/steam2/steamapps/
2 /var/games/steam2/SteamApps/

$  stat -c '%i %n' /var/games/ssd/steamapps/ /var/games/ssd/SteamApps/
2 /var/games/ssd/steamapps/
2 /var/games/ssd/SteamApps/
```

The steamapps/SteamApps directories have the same inode numbers on each dataset.  That's because they're the same directory - there is only one directory for each steam library (it's not a symlink or a hard link, it's one directory that happens to match name lookups in upper & lower case because the fs is case-insensitive).

Coincidentally, they also have the same inode number, 2, on the different datasets...that's because "steamapps" was the first sub-directory created in both the two different datasets, because they were both created to **be** Steam libraries, they have no other purpose.   I'm only mentioning it to highlight the fact that it's not relevant, it's just a coincidence of timing.

But protontricks still complains about them.  I expect it would do the same if the steam libraries were on an NTFS or FAT filesystem (I haven't tested this as I don't have any).

This would also happen with ext4 filesystems on recent kernels (v5.2 or newer) where the "casefold" attribute has been set on a directory (e.g. `chattr +F /path/to/steam/library/`)

**Comment by Matoking** [comment]
Good catch. Considering the efforts to introduce case-insensitivity to ext4 for better Wine support, I should've seen this coming. :)

The most universal check (if not the most performant), is to iterate the list of names in a directory and check if two or more directories match `steamapps` case-insensitively. I've added such a fix in the `case_insensitive_fs` branch. Could you test if it fixes the issue? If you've installed Protontricks using pipx, the command to install it is:

```
$ pipx install --force git+https://github.com/Matoking/protontricks.git@case_insensitive_fs
```

**Comment by craig-sanders** [comment]
> Good catch. Considering the efforts to introduce case-insensitivity to ext4 for better Wine support, I should've seen this coming. 
 :)

yeah well, I don't think too many people use ext4 casefold yet....but WINE and Mono apps were some of the major reasons for it being added.

> The most universal check (if not the most performant), is to iterate the list of names in a directory and check if two or more directories match `steamapps` case-insensitively. I've added such a fix in the `case_insensitive_fs` branch.

I'll probably have time to check this tomorrow.

In the meantime, I've had a quick look at the commit and I'm not sure that this will work as intended.  I may be reading it wrong (because I haven't looked at the protontricks code in any detail) but it seems like it's just checking if there are 2 or more directories matching (case-insensitive) "steamapps".   That would work fine on a case-insensitive fs like my ZFS datasets, but would do the opposite of what's required on a case-sensitive fs that did have two or more such directories.  i.e. it's fixed case-insensitive but broken case-sensitive.

**Comment by Matoking** [comment]
I pushed an update to the branch with a test. I also tested the fix on a NTFS drive mounted using lowntfs-3g with case-insensitive options and it behaves as expected.

The code in question first checks if `steamapps` and `SteamApps` both exist; this can happen if there really are two distinct directories with those names, or if we're in a case-insensitive file system. The commit adds an additional sanity check where we check the number of directory names that match `steamapps` when converted to lowercase. If two or more directories are found, then both `steamapps` and `SteamApps` do exist and we're in a case-sensitive FS, otherwise it's a case-insensitive FS in which case we skip the warning.
